### PR TITLE
adding multi floor support

### PR DESCRIPTION
**Pull request by ppisljar**
resolves https://github.com/Hypfer/Valetudo/issues/69

adds two buttons to the home panel:
- clicking save map asks for a name, then it copies the map to a backup folder with provided name
- clicking load map lists all the saved maps you created and allows you to restore one

it works the way @hagenuck1 described in the original issue. as follow up to this i hope to get it working without the reboot.

it should also work over MQTT custom commands "load_map" and "save_map" but i haven't tested those.

image

### CODE REVIEW

**Comment by hagenuck1** [comment]
Hi, see my updates here:
https://github.com/hagenuck1/Valetudo/wiki/Load-saved-maps-using-Home-Assistant

I found out a reboot isn't needed! But you have to "service rrwatchdoge restart" and wait 30 Seconds for the new map to be loaded. If you start cleaning within this time robot will reboot and may reset!

File: lib/MqttClient.js [hunk 1]
```diff
@@ -419,6 +422,40 @@ MqttClient.prototype.handleCustomCommand = function (message) {
                     console.info("Missing spot_id");
                 }
                 break;
+
+            /**
+             * {
+             *   "command": "go_to",
+             *   "spot_id": "Somewhere"
+             * }
```
Comment (hagenuck1) [inline lib/MqttClient.js]: Comment should be edited:
/**
             * {
             *   "command": "store_map",
             *   "name": "Somemapname"
             * }
             */

File: lib/MqttClient.js [hunk 2]
```diff
@@ -419,6 +422,40 @@ MqttClient.prototype.handleCustomCommand = function (message) {
...
+                    MapManager.storeMap(msg.name, function(err, data) {
+                        if (err) { console.info(err); }
+
+                    })
+                } else {
+                    console.info("Missing name");
+                }
+                break;
+
+            /**
```
Comment (hagenuck1) [inline lib/MqttClient.js]: Comment should be edited:
/**
             * {
             *   "command": "load_map",
             *   "name": "Somemapname"
             * }
             */

File: lib/MqttClient.js [hunk 3]
```diff
@@ -14,7 +15,9 @@ const MQTT_COMMANDS = {
 
 const CUSTOM_COMMANDS = {
     GO_TO: "go_to",
-    ZONED_CLEANUP: "zoned_cleanup"
+    ZONED_CLEANUP: "zoned_cleanup",
+    LOAD_MAP: "load_map",
+    STORE_MAP: "store_map",
```
Comment (hagenuck1) [inline lib/MqttClient.js]: comma after "store map", should be needed?

**Comment by ppisljar** [comment]
thanks for the feedback, i'll update based on your guidelines

i have some quick questions, where answers might help me make this work better:
- can we get the current position ? (x/y coordinates where the robot currently is in the map) 
- and can we change/update the current position ? (start point in your map?)
- is there a way to detect that robot is lifted off the ground ?

**Comment by hagenuck1** [comment]
I haben tried that so far, but the robot will detach its position automatically when the map and the rrwatchdoge are restarted and you start cleaning/ goto. 

I don’t know of any way to get known about lifting up, sorry.

**Comment by ppisljar** [comment]
if robot is lifted and you try to perform an action it signals an error ... possible improvement:

- with each floor set the 'stairs' location, this is the location where robot will move to wait for you to help it up the stairs.
- when requested/needed to move to different floor robot would move to this 'stairs' location
- it would then loop:
  - play a sound to signal it needs help up/down the stairs
  - try to move a bit, if lifted we know someone is helping us
- once we know we are lifted we can assume we are being dropped of at the right location
- loop:
  - try to move: once lifted error is gone we know we are in the new floor
- do the action that was required
- move back to the 'stairs' location in the second floor and again do the 'help me' procedure.

**Comment by ppisljar** [comment]
for initial PR however i think this is fine. rather than above the next improvement would be showing all maps in map view (allowing you to view all saved maps and configure zones for them)

**Comment by hagenuck1** [comment]
Sounds awesome, will build the image and test mqtt tomorrow.

**Comment by ppisljar** [comment]
currently restoring should work:
- save map (Floor1)
- if map is reset load map (Floor1)

it would also be possible to automatically do this if map was reset (set the default floor, on map reset load the default map) but i think that might generate other issues so for now i am leaving it out.

**Comment by hagenuck1** [comment]
Just compiled and tried it. 

Web Interface in Valetudo is working as expected 👍

Mqtt is working fine, too. But errorhandling for non existing floor names is missing and causes the current map to be deleted.

**Comment by ppisljar** [comment]
i will add more error checking

**Comment by vincegio** [comment]
Please add an option in settings to enable/disable these buttons, if this is not done already :)

**Comment by ppisljar** [comment]
this should be ready now

**Comment by hagenuck1** [comment]
Just compiled the new version.

Edit the Multifloor settings and saving it causes valetudo to show "there was an error: 400" to be shown in the bottom.

Store Maps causes it to show an error 500 for me. In the file system it creates an folder with my entered name, but the folder is empty.

Loading maps using the website works fine.

**Comment by vincegio** [comment]
Was curious of the work so booted it up quickly without having my vacuum on the net at the moment.
image
Maybe look into so the buttons are disabled in these states?

**Comment by hagenuck1** [comment]
Thanks, great work! 

Tested and saving & loading map is working fine in web view and mqtt. 

Didn’t found any issues.

**Comment by ppisljar** [comment]
anything to be done before it can be merged ?

**Comment by hagenuck1** [comment]
Didn’t got any feedback and haven’t found more issues, I think it could be merged.

**Comment by pszafer** [comment]
@ppisljar maybe you could add status to mqtt which map/floor is currently loaded.

**Comment by ppisljar** [comment]
@pszafer i can do that in the next PR, first i'll wait for this to get merged.

**Comment by Hypfer** [comment]
I don't trust it. Provide proof that it will work reliably (long term!) or cease.

**Comment by iMicknl** [comment]
@Hypfer is it an idea to have an experimental branch + build for those kind of features?

**Comment by TheRealMoeder** [comment]
I can see the issue you have with this. Maybe an experimental branch with builds will indeed make testing more widespread.

**Comment by hagenuck1** [comment]
I also can fully understand you Hypfer and know that the unknown factory resets should be fixed before getting more potential crashing issues. 

But just to be clear, the RoboController.cfg isn’t needed to be changed for map reloading? Only thing that’s needed to reload the map is restarting the rrwatchdoge to reload the map files. 

For me the loading of the map files using manual scripts and now this pr is working fine since about 2019-06-06 and the robot hadn’t any resets or problems till now. 

Also it survived the 8 week reset and runs fine since 9 weeks now. (Recreated the image to use a new ssh key) 

My map files changing date is 2019-07-07, so they’re working longer as 6/8 weeks as somebody mentioned somewhere.

**Comment by ppisljar** [comment]
i also didn't have any issues so far, the robot works as expected. Also not sure what core functionality am i messing with here :| . i also agree that there should be a development branch and stable one. No idea how i can provide the proof but if this contribution is unwanted i also have other things to do :)

**Comment by Hypfer** [comment]
Well that turned out wrong.
Let me give you the full version of what I was trying to say further up in this thread:

`I don't` think that just looking at the Code to verify that everything is working will be enough here, due to the fact, that there is another Component involved which doesn't have a well-defined interface nor sourcecode available to check the behaviour. If you mention me to get me to merge this PR, please   `provide proof that it will work reliably (long term!) or cease` mentioning me altogether.

I just hate emails like that. Still I should've been more clear there.

Questions/ that would need to be resolved before this can be merged:
- Do we know for certain that simply swapping the data files and restarting the cleaner process is all we need to do to make multiple maps work? Where did that information come from?
- Are we sure that there is no hidden reset counter somewhere which increases each time the process is respawned?
- Does it work when used daily? We got at least one report here from hagenuck1 so thats something
- What about Gen1 vacuums?
- ~~What about the overall code quality~~ ✔️

There's also a strategic decision to be made here: Should Valetudo stick to (semi-) official APIs, or should it use other means to get things done. What about portability to other Vacuums at some point?

**Comment by hagenuck1** [comment]
More info by me: 
Where did the information come from? 
- I’ve read about the last_map replacement in the Valetudo group and figured out by testing, that replace it (NOT during cleaning!) and then restart rrwatchdoge (and not the whole robot as written in the group is enough) 

Hidden counter: 
- can’t provide there’s no hidden counter unfortunately 

Daily usage: 
- no problems for me 
- see the home assistant 0.100 release notes, there’s somebody who has written a tutorial using my wiki fork which is the same as this PR does. He resets the map every time the robot hits docked. 
— https://twitter.com/oischinger/status/1178254181221769216?s=21
- maybe we could let them report issues somewhere if they find problems, too. 

Gen1
- as they don’t have persistent Maps this shouldn’t work. I think for them the multiple maps feature is inactive.

**Comment by oischinger** [comment]
Hi,
I use a gen1 robot. I didn't use the firmware but I restore via ssh as per instructions. The restoring works fine for me.
I was brave enough to trigger the map reset multiple times in a row in a few seconds and didn't see a reset or reboot.

I'll do some more experiments soon on a firmware including this PR.

**Comment by pax2011** [comment]
> Hi,
> I use a gen1 robot. I didn't use the firmware but I restore via ssh as per instructions. The restoring works fine for me.
> I was brave enough to trigger the map reset multiple times in a row in a few seconds and didn't see a reset or reboot.
> 
> I'll do some more experiments soon on a firmware including this PR.

Do you restore maps in which moment? before start cleaning? Do you use homeassistant to restore it?

**Comment by oischinger** [comment]
I set up an automation that resets the map whenever the robot is docked. Because of these reasons:
a) my Gen1 robot sometimes rotates the map by 90deg which brakes zoned cleanup 
b) I have multiple floors and sometimes carry the robot to another floor. The charger is in the ground floor. I want to restore the map of this floor automatically.

But I also triggered the restore automation multiple times in a row (when it was not cleaning). The restore still worked as expected.

**Comment by pax2011** [comment]
> I set up an automation that resets the map whenever the robot is docked. Because of these reasons:
> a) my Gen1 robot sometimes rotates the map by 90deg which brakes zoned cleanup
> b) I have multiple floors and sometimes carry the robot to another floor. The charger is in the ground floor. I want to restore the map of this floor automatically.
> 
> But I also triggered the restore automation multiple times in a row (when it was not cleaning). The restore still worked as expected.

I set up everything, and it works locally on robot. If I try to run as a script on Hass,io i get 

> Error running command: `ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -i /config/ssh/id_rsa root@192.168.1.220 "bash ../mnt/data/valetudo/maps/floor1.sh"`, return code: 255
NoneType: None

This is my scipt https://ibb.co/bXHpnwH
This is my config https://it.imgbb.com/

What am I missing?

**Comment by pax2011** [comment]
> I set up an automation that resets the map whenever the robot is docked. Because of these reasons:
> a) my Gen1 robot sometimes rotates the map by 90deg which brakes zoned cleanup
> b) I have multiple floors and sometimes carry the robot to another floor. The charger is in the ground floor. I want to restore the map of this floor automatically.
> 
> But I also triggered the restore automation multiple times in a row (when it was not cleaning). The restore still worked as expected.

Now I would like to add another floor with different zones, zones are stored in  /mnt/data/valetudo/config.json. So I'm trying a script thats overwrite config.json and last_map to get 2 floor with differents zones.
It works, but to have correct zones i have to reboot the vacuum. The file is correct but valetudo's interface shows wrongs zones.

SH scipt
`rsync -r ../../../../mnt/data/valetudo/maps/floor1/ ../../../../mnt/data/rockrobo/
rsync -r ../../../../mnt/data/valetudo/maps/jsonfloor1/ ../../../../mnt/data/valetudo/
rm -f ../../../../mnt/data/rockrobo/user_map0
rm -f ../../../../mnt/data/rockrobo/StartPos.data
service rrwatchdoge restart`
Than reboot.

any solution to reload the config.json without rebooting?

**Comment by Hypfer** [comment]
@pax2011 uh.. restart valetudo?
But that will cause the msgIDs to get out of sync which you also don't want.

The config file is read only once on startup. Then, everything happens in-memory. If changes occur, the whole in-memory config will be written to disk.
This can be seen here:
https://github.com/Hypfer/Valetudo/blob/master/lib/Configuration.js

Since the code is open source after all, I'd suggest working with it instead of around it.

**Comment by pax2011** [comment]
> @pax2011 uh.. restart valetudo?
> But that will cause the msgIDs to get out of sync which you also don't want.
> 
> The config file is read only once on startup. Then, everything happens in-memory. If changes occur, the whole in-memory config will be written to disk.
> This can be seen here:
> https://github.com/Hypfer/Valetudo/blob/master/lib/Configuration.js
> 
> Since the code is open source after all, I'd suggest working with it instead of around it.

thanks for your advice, but I really don't know how to do it. I don't have much experience with coding.

The zones are stored in config.json can't get other solution to reload it.
I also thought about doing this:
`killall miio_client && service valetudo restart`
But it would be the same thing, wouldn't it?

**Comment by hagenuck1** [comment]
> I set up an automation that resets the map whenever the robot is docked. Because of these reasons:
> a) my Gen1 robot sometimes rotates the map by 90deg which brakes zoned cleanup
> b) I have multiple floors and sometimes carry the robot to another floor. The charger is in the ground floor. I want to restore the map of this floor automatically.
> 
> But I also triggered the restore automation multiple times in a row (when it was not cleaning). The restore still worked as expected.

Okay, sounds great, that it's working fine for gen1, too!
I hat the 90deg rotation, too on my gen2 before the persistent maps, I think they're caused by starting a full clean by clicking "start". 
- Solution for this should be to create a zone, that contains the whole floor.
Would be great to hear if the PR is also working fine on gen1.

@pax2011 just create all zones directly in valetudo by loading the map first and add a leading name (for example "floor1_") to each zone containing a hint to the current floor. Then repeat this for the other floor(s). This way you could still differentiate between the rooms and have all zones in one config.json.

**Comment by jonathan-reisdorf** [comment]
Nice! :+1: These "experimental" features could also be put into a separate section in the UI to indicate that it's potentially more dangerous to use them, e.g. adding a new card in settings, or combine it with the persistent data feature ("Map settings"?). (Also one could argue that swapping the floor is something between a setting and an action, so it wouldn't be wrong in settings imo.)

**Comment by oischinger** [comment]
My robot recently did a factory reset but it seems it was not related to the map restoring. It reset itself overnight, around 10 hours after the last map restore.

File: lib/MapManager.js [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    if (!fs.existsSync("/mnt/data/rockrobo/last_map")) {
+        return cb("map doesn't exist");
+    }
+    fs.copyFileSync("/mnt/data/rockrobo/last_map", "/mnt/data/valetudo/maps/" + name + "/last_map");
+    if (fs.existsSync("/mnt/data/rockrobo/ChargerPos.data")) fs.copyFileSync("/mnt/data/rockrobo/ChargerPos.data", "/mnt/data/valetudo/maps/" + name + "/ChargerPos.data");
+    else {
+        // we don't have charger info
+    }
+    if (fs.existsSync("/mnt/data/rockrobo/PersistData_1.data")) fs.copyFileSync("/mnt/data/rockrobo/PersistData_1.data", "/mnt/data/valetudo/maps/" + name + "/PersistData_1.data");
+    if (fs.existsSync("/mnt/data/rockrobo/PersistData_2.data")) fs.copyFileSync("/mnt/data/rockrobo/PersistData_1.data", "/mnt/data/valetudo/maps/" + name + "/PersistData_2.data");
```
Comment (N1k145) [inline lib/MapManager.js]: PersistData_1 is copied twice.
Is this correct?

```suggestion
    if (fs.existsSync("/mnt/data/rockrobo/PersistData_2.data")) fs.copyFileSync("/mnt/data/rockrobo/PersistData_2.data", "/mnt/data/valetudo/maps/" + name + "/PersistData_2.data");
```

**Comment by BoukeHaarsma23** [comment]
[Copy from from #91]
Just wanted to add something here. Apparently there is a beta with multi-room support floating around.

See [the announcement on facebook.](https://www.facebook.com/groups/1963580733664463/permalink/2865795586776302/)

If somebody has there hands on it. Maybe it is possible to use this in valetudo as well?

**Comment by jonathan-reisdorf** [comment]
The new firmware 2008 is linked on vacuumz.info and can be downloaded directly from the xiaomi CDN by assembling the base URL `https://cdn.awsbj0.fds.api.mi-img.com/rubys/updpkg/` and the `FileName` displayed on said website for FW 2008.

**Comment by jonathan-reisdorf** [comment]
Didn't test it yet but the announcements on facebook sound promising (also adds automatic room identification and selective room cleaning)

**Comment by hagenuck1** [comment]
It’s only multi room support and not multi floor, as I understood.

**Comment by jonathan-reisdorf** [comment]
Thanks for clarifying. Then it's not related to this PR. (Even though it would be great if somebody tested the new FW and reported if it works with Valetudo, and, if it's still possible to factory reset to original FW to regain rootability)

**Comment by arieltools** [comment]
As I was reading Hypfer's concern [here](https://github.com/Hypfer/Valetudo/pull/317#issuecomment-540929686) about portability to other vacuums, it could also be considered to group these "advanced features" into "add-ons" or "plugins". This will allow to set a "compatibility flag" for each add-on, effectively grouping these per-vacuum in a future multi-vacuum release. 

Then, Valetudo could stick to the [semi-]official APIs and to develop the core further, while allowing for more advanced features to be developed in parallel by the core team or other interested parties.

> Nice!  These "experimental" features could also be put into a separate section in the UI to indicate that it's potentially more dangerous to use them, e.g. adding a new card in settings, or combine it with the persistent data feature ("Map settings"?). (Also one could argue that swapping the floor/map is something between a setting and an action, so it wouldn't be wrong in settings imo. On the other hand, so is changing the fan speed...)

**Comment by white-gecko** [comment]
Why is this closed? Is there a solution for the multi floor issue?

**Comment by alexkn** [comment]
There is no progress here.
#69 is still open, if someone want to implement it in an good way.

**Pull request closed without merging**